srml/staking/src/slashing.rs

```diff
@@ -0,0 +1,772 @@
+pub(crate) fn slash<T: Trait>(params: SlashParams<T>) -> RewardPayout<T> {
+	let SlashParams {
+		stash,
+		slash,
+		exposure,
+		slash_era,
+		window_start,
+		now,
+		reward_proportion,
+	} = params.clone();
+
+	let mut reward_payout = Zero::zero();
+	let mut slash_imbalance = <NegativeImbalanceOf<T>>::zero();
+
+	// is the slash amount here a maximum for the era?
+	let own_slash = slash * exposure.own;
+	let (prior_slash_p, era_slash) = <Module<T> as Store>::ValidatorSlashInEra::get(
+		&slash_era,
+		stash,
+	).unwrap_or((Perbill::zero(), Zero::zero()));
+
+	if own_slash > era_slash {
+		<Module<T> as Store>::ValidatorSlashInEra::insert(
+			&slash_era,
+			stash,
+			&(slash, own_slash),
+		);
+	} else {
+		// we slash based on the max in era - this new event is not the max,
+		// so neither the validator or any nominators will need an update.
+		//
+		// this does lead to a divergence of our system from the paper, which
+		// pays out some reward even if the latest report is not max-in-era.
+		// we opt to avoid the nominator lookups and edits and leave more rewards
+		// for more drastic misbehavior.
+		return RewardPayout(None)
+	}
+
+	// apply slash to validator.
+	{
+		let mut spans = fetch_spans::<T>(
+			stash,
+			window_start,
+			&mut reward_payout,
+			&mut slash_imbalance,
+			reward_proportion,
+		);
+
+		let target_span = spans.compare_and_update_span_slash(
+			slash_era,
+			own_slash,
+		);
+
+		if target_span == Some(spans.span_index()) {
+			// misbehavior occurred within the current slashing span - take appropriate
+			// actions.
+
+			// chill the validator - it misbehaved in the current span and should
+			// not continue in the next election. also end the slashing span.
+			spans.end_span(now);
+			<Module<T>>::chill_stash(stash);
+
+			// make sure to disable validator till the end of this session
+			if T::SessionInterface::disable_validator(stash).unwrap_or(false) {
+				// force a new era, to select a new validator set
+				crate::ForceEra::put(crate::Forcing::ForceNew);
+			}
+		}
+	}
+
+	reward_payout += slash_nominators::<T>(params, prior_slash_p, &mut slash_imbalance);
+
+	RewardPayout(Some((reward_payout, slash_imbalance)))
+}
+
+/// Slash nominators. Accepts general parameters and the prior slash percentage of the nominator.
+///
+/// Returns the amount of reward to pay out.
+fn slash_nominators<T: Trait>(
+	params: SlashParams<T>,
+	prior_slash_p: Perbill,
+	slash_imbalance: &mut NegativeImbalanceOf<T>,
+) -> BalanceOf<T> {
+	let SlashParams {
+		stash: _,
+		slash,
+		exposure,
+		slash_era,
+		window_start,
+		now,
+		reward_proportion,
+	} = params;
+
+	let mut reward_payout = Zero::zero();
+
+	for nominator in &exposure.others {
+		let stash = &nominator.who;
+
+		// the era slash of a nominator always grows, if the validator
+		// had a new max slash for the era.
+		let era_slash = {
+			let own_slash_prior = prior_slash_p * nominator.value;
+			let own_slash_by_validator = slash * nominator.value;
```

Comment:
Why do we multiple `slash` value and the exposed balance? Shouldn't this be just a percentage?

Also shouldn't we subtract validators `own_slash` from the remaining `slash` (before calling this function) and only slash nominators the rest?

Reply:
> Also shouldn't we subtract validators own_slash from the remaining slash (before calling this function) and only slash nominators the rest?

There is a way of doing that, but it's really involved and the Web3 Research team mentioned that we do in fact want slashes to hit nominators at least a little so that they have a disincentive to nominate untrusted candidates.

Reply:
`slash` is a proportion here, also